api/src/pcapi/core/offers/api.py

```diff
@@ -1012,16 +1012,24 @@ def add_criteria_to_offers(
     if not offer_ids:
         return False
 
+    # check existing tags on selected offers to avoid duplicate association which would violate Unique constraint
+    existing_criteria_on_offers = (
+        criteria_models.OfferCriterion.query.filter(criteria_models.OfferCriterion.offerId.in_(offer_ids))
+        .with_entities(criteria_models.OfferCriterion.offerId, criteria_models.OfferCriterion.criterionId)
+        .all()
+    )
+
```

Comment:
Je ne me rends pas compte du volume des tags (OfferCriterion), mais si on risque d'en récupérer beaucoup, ça vaut peut-être le coup de construire un `set` de `tuple` au lieu d'une `liste` (de `tuple`). Mais si ça reste raisonnable, c'est une fausse optimisation inutile. Encore une fois, je ne me rends pas compte du volume que ça peut représenter.

Reply:
@jeremieb-pass : comme il y a unicité sur les tuples, un set de tuples renverra exactement le même nombre de lignes qu'une liste de tuples.
Au départ j'ai fait un premier essai pour vérifier le principe, en vérifiant unitairement l'existence d'un tuple dans la table, à chaque itération de la double boucle. Je pense qu'une requête unique est plus optimale quand même :)

Reply:
ça ne me parait pas être une optimisation super utile. d'autant que ça peut éventuellement nous faire économiser de la mémoire, mais peut être perdre du temps de traitement à la construction du `set`.